### PR TITLE
Create s3-private module and pvsite ml bucket

### PR DESCRIPTION
**Pull request by devsjc**
Genericise s3 module and update deprecated resources

### CODE REVIEW

File: terraform/modules/storage/s3-private/variables.tf [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+variable "service_name" {
+  description = "The name of the service for which the bucket supplies"
+  type = string
+}
+
+variable "lifecycled_prefixes" {
+  description = "The data-slash prefixes requiring lifecycles"
+  type = list(string)
```
Comment (peterdudfield) [inline terraform/modules/storage/s3-private/variables.tf:23]: can you have the option of none? If we want to keep some buckets all the time

File: terraform/modules/storage/s3-private/variables.tf [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+variable "domain" {
+  description = "The domain of the S3 bucket"
+  type = string
+}
+
+variable "service_name" {
+  description = "The name of the service for which the bucket supplies"
+  type = string
+}
+
```
Comment (peterdudfield) [inline terraform/modules/storage/s3-private/variables.tf:20]: do we have an option of how long each bucket will keeps its data?

File: terraform/modules/storage/s3-private/variables.tf (hunk 1 above)
Comment (devsjc) [reply terraform/modules/storage/s3-private/variables.tf:23]: Yes, an empty list does that [], like in the case for the s3-ml one in main

File: terraform/modules/storage/s3-private/bucket.tf [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+  bucket = aws_s3_bucket.bucket.id
+  rule {
+    id = "remove_old_${each.key}_files"
+    status = "Enabled"
+    filter {
+      prefix = "${each.key}/"
+    }
+    expiration {
+      days = 7
```
[the remark this answers is not in the document]
Comment (devsjc) [reply terraform/modules/storage/s3-private/bucket.tf:46]: It only creates these 7 day lifecycles for the specified prefixes - this mirrors the setup prior to this refactor: 

```
# Documentation: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket
resource "aws_s3_bucket" "s3-nwp-bucket" {
  bucket = "nowcasting-nwp-${var.environment}"
  acl    = "private"

  lifecycle_rule {
    id      = "remove_old_files"
    enabled = true

    prefix = "data/"

    expiration {
      days = 7
    }
  }

  lifecycle_rule {
    id      = "remove_old_raw_files"
    enabled = true

    prefix = "raw/"

    expiration {
      days = 7
    }
  }

  tags = {
    Name        = "${var.environment}-s3"
    Environment = "${var.environment}"
  }
}
```

File: terraform/modules/storage/s3-private/variables.tf (hunk 1 above)
Comment (devsjc) [reply terraform/modules/storage/s3-private/variables.tf:23]: Also, I think this only setting lifecycles on prefixes within the bucket, as opposed to on the entire bucket

**Comment by devsjc** [comment]
I have removed all changes to the nowcasting side of things, moved the changes into their own module that PVSite uses to create its ml bucket, left s3 to call the old module so no buckets will be deleted over there

**Comment by peterdudfield** [comment]
this should do this one - https://github.com/openclimatefix/ocf-infrastructure/issues/155?

**Pull request merged**